FreeBSD packages: Init at 13.1

Ericson2314 opened the pull request:
###### Motivation for this change

Have working
```sh
nix-build -A pkgsCross.x86_64-freebsd.hello
nix-build -A pkgsCross.x86_64-freebsd.zlib
nix-build -A pkgsCross.x86_64-freebsd.zstd
nix-build -A pkgsCross.x86_64-freebsd.freebsd.sys
```

~~Running out of memory as NetBSD `makeMinimal` fails to build. Weird, know what's up @matthewbauer?~~

Oh netbsd make is cached, duh, this is freebsd libc itself.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC @wahjava

Ericson2314 committed:
freebsd packages: Init at 13.1

Always set `SRCTOP`, set it with abs path

llvmPackages: Bump minimum version for FreeBSD

llvmPackages_*, libgcc, compiler_rt: Hack in enough libs that one can compiler C

freebsd.compat: Rename some things to work around cc-wrapper change

0bea4a194f7103fcedc0226a74e50657875a7c4d / #191724 in particular

vcunat commented on lib/systems/examples.nix:
```diff
@@ -228,11 +228,26 @@ rec {
+  };
+
+  # Deprecate
+  amd66-netbsd = x86_64-netbsd;
```
Should this be `amd64-netbsd`?

Ericson2314 replied:
Haha yes

wahjava commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -0,0 +1,69 @@
+  version = "12";
+
+  fetchFreeBSD = path: sha256: fetchsvn {
+    url = "https://svn.freebsd.org/base/stable/${version}/${path}";
```
`stable` branch [may not be](https://www.freebsd.org/doc/en_US.ISO8859-1/books/handbook/current-stable.html) stable. You can use [one of the releng branch](https://svnweb.freebsd.org/base/releng/12.1/) to track released code. Also you might want to stick to a particular svn revision, and not HEAD

Ericson2314 replied:
Thanks!

Ericson2314 commented:
I dropped link to this in https://forums.freebsd.org/threads/can-i-compile-a-freebsd-kernel-from-linux.64582/, a that is vaguely relevant.

Ericson2314 commented:
OK need to fix makefiles or bmake itself so non-default system rules `-m` propagate to recursive make---probably easiest thing to add an environment variable to bmake for this.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10496,6 +10490,17 @@ in
+    if stdenv.targetPlatform.isi686
+      then gcc6.cc
+    else if stdenv.targetPlatform == stdenv.hostPlatform && targetPackages.stdenv.cc.isGNU
+	  # Can only do this is in the native case, otherwise we might get infinite
+	  # recursion if `targetPackages.stdenv.cc.cc` itself uses `gccForLibs`.
```
Mixing tabs and spaces.

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -0,0 +1,287 @@
+{ stdenv, lib, stdenvNoCC
+, pkgsBuildBuild, pkgsBuildHost, pkgsBuildTarget, pkgsHostHost, pkgsHostTarget, pkgsTargetTarget
+, buildPackages
+, bsdSetupHook, fetchsvn, fetchurl, groff, mandoc, yacc, flex
```
yacc is bison now.

alyssais commented:
Let's maybe change the title of this now, since NetBSD cross is already fixed?

alyssais commented:
This is also going to have to target staging because of all the NetBSD rebuilds on Darwin.

sternenseemann commented on pkgs/os-specific/bsd/netbsd/default.nix:
```diff
@@ -38,12 +29,12 @@ let
       bsdSetupHook netbsdSetupHook
-      makeMinimal install tsort lorder mandoc groff statHook
+      makeMinimal
+      install tsort lorder mandoc groff statHook
```
A bit of a nitpick, but this is a bit of a scoping mess with thing explicitly coming from `buildPackages.netbsd` and stuff like `mandoc` coming from the expressions inputs (and thus implicitly from `buildPackages`).

 This is mostly about readabilty, I was a bit confused about this for half a minute.

Also applies to FreeBSD's `mkDerivation`.

sternenseemann commented on pkgs/os-specific/bsd/netbsd/default.nix:
```diff
@@ -453,39 +447,35 @@ let
-    # meta.platforms = lib.platforms.netbsd;
-    makeFlags = [ "RPCGEN_CPP=${buildPackages.gcc-unwrapped}/bin/cpp" ];
+    meta.platforms = lib.platforms.netbsd;
+    makeFlags = [ "RPCGEN_CPP=${buildPackages.stdenv.cc.cc}/bin/cpp" ];
```
Wouldn't just `stdenv` be sufficient or correcter here? Or do buildPackages.stdenv.cc.cc and stdenv.cc.cc have the same effective hostPlatform they run on?

Ericson2314 replied:
At this point we don't yet have a CPP for the right stage, so we awkwardly use the previous stage's.

sternenseemann replied:
I see, thanks for the clarification.

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -0,0 +1,347 @@
+      mv "$BSDSRCDIR/${path}/lorder.sh" "$out/bin/lorder"
+      chmod +x "$out/bin/lorder"
+      mv "$BSDSRCDIR/${path}/lorder.1" "$man/share/man"
+      exit 0;
```
?

Ericson2314 replied:
Oh this is WIP as hell and full of old garbage and half-done rebases.

Ericson2314 replied:
Fixed.

L-as commented:
What is blocking this?

Ericson2314 commented:
I suppose we could merge soon, but it doesn't yet work. See netbsd. There are a lot of things to package before one gets to hello world.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/status-of-freebsd/17786/4

sash-kan commented:
i tried to build "hello" from this branch: https://github.com/Ericson2314/nixpkgs/tree/bsd-cross
and got an error from bmake:
```
$ nix-build . -A pkgsCross.x86_64-freebsd.hello
...
make[1]: "/nix/store/j4z7lyxrn29lzibfykg24d2plfwq1nfg-bmake-freebsd-13.0.0/share/mk/src.libnames.mk" line 701: /build/libc-freebsd-filtered-src/lib/msun: Missing or incorrect value for LIBMDIR in src.libnames.mk: lib/msun
make: "/build/libc-freebsd-filtered-src/lib/libc/Makefile" line 197: warning: "make MK_MAKE_CHECK_USE_SANDBOX=yes -B -C ../../lib/msun -V ARCH_SUBDIR" returned non-zero status
make: make: don't know how to make asctime.c. Stop
make: stopped in /build/libc-freebsd-filtered-src/lib/libc
error: builder for '/nix/store/w5z1a58lcd23miyjw5z7lrg6phplrp88-libc-x86_64-unknown-freebsd-13.0.0.drv' failed with exit code 2;
```

sash-kan commented:
the problem is that the [variable `SRCTOP`](https://github.com/Ericson2314/nixpkgs/blob/0c7af40f0d3663ccf6c6bb1cf6d9ba7cb51c5c25/pkgs/os-specific/bsd/freebsd/default.nix#L518) is substituted only when calling bmake for [`path`](https://github.com/Ericson2314/nixpkgs/blob/0c7af40f0d3663ccf6c6bb1cf6d9ba7cb51c5c25/pkgs/os-specific/bsd/freebsd/default.nix#L503).
and for [`extrapaths`](https://github.com/Ericson2314/nixpkgs/blob/0c7af40f0d3663ccf6c6bb1cf6d9ba7cb51c5c25/pkgs/os-specific/bsd/freebsd/default.nix#L504-L511) it remains empty.

Ericson2314 commented:
@sash-kan thanks for looking into this!!

Ericson2314 commented:
current error:
```
rpc_scan.c: In function 'get_token':
rpc_scan.c:169:66: error: 'WUNTRACED' undeclared (first use in this function)
  169 |                                         waitpid(childpid, &stat, WUNTRACED);
      |                                                                  ^~~~~~~~~
```
I don't get this because if I go in a nix shell `WUNTRACED` is defined:
```bash
nix-shell . -A pkgsCross.x86_64-freebsd.buildPackages.freebsd.rpcgen
```
```bash
echo -e '#include <sys/wait.h>\nWUNTRACED' | $CC -E -x c - | tail
```

Ericson2314 commented:
current error:
```
rpc_scan.c: In function 'get_token':
rpc_scan.c:169:66: error: 'WUNTRACED' undeclared (first use in this function)
  169 |                                         waitpid(childpid, &stat, WUNTRACED);
      |                                                                  ^~~~~~~~~
```
I don't get this because if I go in a nix shell `WUNTRACED` is defined:
```bash
nix-shell . -A pkgsCross.x86_64-freebsd.buildPackages.freebsd.rpcgen
```
```bash
echo -e '#include <sys/wait.h>\nWUNTRACED' | $CC -E -x c - | tail
```

Ericson2314 commented:
Next problem: I think the headers need to be built separately, like NetBSD.

Ericson2314 commented:
I can build hello and zlib now!!!

Ericson2314 commented:
It's messy (e.g. need to sync changes between LLVM versions) but it works! Have at it, y'all!

Ericson2314 commented:
Temporary changing the target branch so we get smaller change until `staging` hits `master`.

Ericson2314 commented:
The kernel builds!

SuperSandro2000 commented on pkgs/os-specific/bsd/xargs-j.sh:
```diff
@@ -0,0 +1,17 @@
+
+while (( $# )); do
+    if [[ "$1" = "$token" ]]; then
+		mapfile -t -O $(("${#args[@]}" + 1)) args
```
You are mixing tabs and spaces here

SuperSandro2000 commented on pkgs/os-specific/bsd/xargs-j.sh:
```diff
@@ -0,0 +1,17 @@
+    shift
+done
+
+"${args[@]}"
```
```suggestion
exec "${args[@]}"
```

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -120,6 +120,9 @@ in lib.makeScopeWithSplicing
     HOST_SH = stdenv'.shell;
 
+    # Since STRIP below is the flag
+    STRIPBIN="${stdenv.cc.bintools.targetPrefix}strip";
```
```suggestion
    STRIPBIN = "${stdenv.cc.bintools.targetPrefix}strip";
```

pbsds commented:
I tried building `github:Ericson2314/nixpkgs/bsd-cross#pkgsCross.x86_64-freebsd.hello` on a arch machine without sandboxing enabled, and got some _interesting_ output:

```
...
compat> rm: cannot remove '/usr/bin/ppmtopjxl': Permission denied
compat> rm: cannot remove '/usr/bin/nvidia-cuda-mps-control': Permission denied
compat> rm: cannot remove '/usr/bin/weston-simple-shm': Permission denied
compat> rm: cannot remove '/usr/bin/pamdice': Permission denied
compat> rm: cannot remove '/usr/bin/eqn2graph': Permission denied
compat> rm: cannot remove '/usr/bin/curl': Permission denied
compat> rm: cannot remove '/usr/bin/gsdj': Permission denied
...
```

It seems like it tried to clean the `/usr/bin` directory on host the machine

SuperSandro2000 commented:
> on a arch machine without sandboxing enabled, and got some _interesting_ output:

Is this required for *bsd's somehow because at this point we mostly ignore errors without sandboxing enabled.

7c6f434c commented on pkgs/development/compilers/llvm/12/compiler-rt/default.nix:
```diff
@@ -91,6 +93,8 @@ stdenv.mkDerivation {
     ln -s $out/lib/*/clang_rt.crtend-*.o $out/lib/crtend.o
     ln -s $out/lib/*/clang_rt.crtbegin_shared-*.o $out/lib/crtbeginS.o
     ln -s $out/lib/*/clang_rt.crtend_shared-*.o $out/lib/crtendS.o
+  '' + lib.optionalString (doFakeLibgcc) ''
+    ln -s $out/lib/freebsd/libclang_rt.builtins-*.a $out/lib/libgcc.a
```
Hm, does that mean `doFakeLibgcc` only works on FreeBSD?

Ericson2314 replied:
It is just used for FreeBSD right now. I wish we didn't have to do this but the FreeBSD Clang frontend just has these sorts of things hard-coded because that's how FreeBSD currently bootstrap things.

The larger issue is the "OS" field of the target name really has nebulous meaning (is it the syscall ABI, the libc ABI, something more? etc.) and so we get these postels law gone wrong situations of people hacking in ad-hoc special cases to get things working that just cause extra headache for something like Nixpkgs where we rather do all the combinations as uniformly as possible.

9glenda commented:
Great work

9glenda commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -0,0 +1,900 @@
+
+    while (( $# )); do
+      if (( $# == 1 )); then
+        if (( $path_args > 1)) || [[ "$1" = */ ]]; then
```
In bash the first `$` is not needed here.

9glenda replied:
If this is supposed to be a posix shell script (I assume it's bash) it is not posix compliant

9glenda commented on pkgs/os-specific/bsd/freebsd/default.nix:
```diff
@@ -0,0 +1,900 @@
+    '';
+
+    postInstall = lib.optionalString (!stdenv.targetPlatform.isFreeBSD) ''
+      boot_mk=$BSDSRCDIR/tools/build/mk
+      cp $boot_mk/Makefile.boot* $out/share/mk
+      replaced_mk=$out/share/mk.orig
+      mkdir $replaced_mk
+      mv $out/share/mk/bsd.{lib,prog}.mk $replaced_mk
+      for m in bsd.{lib,prog}.mk; do
+        cp $boot_mk/$m $out/share/mk
+        substituteInPlace $out/share/mk/$m --replace '../../../share/mk' '../mk.orig'
+      done
```
Generally you should double quote (`"`) variables to prevent word splitting and globbing

9glenda reviewed (commented):
I took a quick look at the shell code. It's nothing that bad just really small "mistakes"

9glenda commented:
https://github.com/Ericson2314/nixpkgs/pull/1
I implemented the changes myself and opened a PR

Ericson2314 commented:
OK I synced the LLVM changes across all the versions. I think this is basically ready to go!

Ericson2314 committed:
lib/systems: Simplify NetBSD examples

The libc will be inferred.

Ericson2314 committed:
lib/systems: Support FreeBSD

A tricky thing about FreeBSD is that there is no stable ABI across
versions. That means that putting in the version as part of the config
string is paramount.

We have a parsed represenation that separates name versus version to
accomplish this. We include FreeBSD versions 12 and 13 to demonstrate
how it works.

sternenseemann commented on lib/systems/parse.nix:
```diff
@@ -290,7 +290,11 @@ rec {
+    # versions. That means that putting in the version as part of the
+    # config string is paramount.
+    freebsd12 = { execFormat = elf;     families = { inherit bsd; }; name = "freebsd"; version = 12; };
+    freebsd13 = { execFormat = elf;     families = { inherit bsd; }; name = "freebsd"; version = 13; };
```
IIRC Nix's configure script makes a point of stripping version information from system doubles, so it might be impossible to get such a double as `builtins.currentSystem`.

Ericson2314 replied:
You and @alyssais should work that out :) I am not sure what is more correct. If what we have is more correct, that is fine with me we can worry about native compilation later! :D

kirillrdy commented:
Hi, firstly thank you for all the effort that was put into this MR !

unfortunately when i try

`nix-build -A pkgsCross.x86_64-freebsd.hello`

I get 

```
building '/nix/store/ffa2r42x74j5mfsvkymj9s74w114kqyr-stdenv-linux.drv'...
remote: Enumerating objects: 91213, done.
remote: Counting objects: 100% (91213/91213), done.
remote: Compressing objects: 100% (78514/78514), done.
error: RPC failed; curl 56 OpenSSL SSL_read: error:0A000119:SSL routines::decryption failed or bad record mac, errno 0
error: 5347 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
remote: Enumerating objects: 91213, done.
remote: Counting objects: 100% (91213/91213), done.
remote: Compressing objects: 100% (78514/78514), done.
```

can someone double check that this builds for others ? just so I know if there is something with my setup

kirillrdy commented:
hi, actually i found the answer,  something is broken with ssl on master, i've rebuild my nixos from previous well known configuration and now it seems to be building

Mic92 commented:
Is c++ known to work? I just tried cross-compiling patchelf in the patchelf repo and got this:

```
x86_64-unknown-freebsd13-ld: error: undefined symbol: _Unwind_Resume
>>> referenced by patchelf.cc:0
>>>               patchelf.o:(getElfType(std::__1::shared_ptr<std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > const&))
>>> referenced by patchelf.cc:0
>>>               patchelf.o:(error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&))
>>> referenced by new:0 (/nix/store/qbbik1n6qmmbwhc74fv9czm8qy21cmvf-libcxx-x86_64-unknown-freebsd13-12.0.1-dev/include/c++/v1/new:0)
>>>               patchelf.o:(resolveArgument(char const*))
>>> referenced 81 more times
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```

9glenda commented:
Finally this is merged

Ericson2314 commented:
@Mic92 I built zstd and it worked so some C++ does, I thought.

9glenda commented:
This pr just adds cross compilation support right? Or does it also add support for building the FreeBSD kernel on FreeBSD in case you get nix running there?

Ericson2314 commented:
@9glenda only cross compilation support, but I did cross compile the kernel too!

Ericson2314 commented:
My next stop would try to be NetBSD and FreeBSD in https://github.com/cleverca22/not-os. I think just using Nix on FreeBSD will actually be harder because modern FreeBSD once to mandate that all syscalls go through the same userland libc --- either we rebuild the entire userland with Nix, or we do impurities.

Mic92 commented:
It seems that with certain c++ features (i.e. c++17), you need libunwind linked as well.

This is what I added for patchelf to get it working:

```
  buildInputs = [ llvmPackages.libunwind ];
  NIX_LDFLAGS = "-lunwind";
```

Probably the macOS version already has something implemented?

Ericson2314 commented on pkgs/development/compilers/llvm/10/default.nix:
```diff
@@ -156,7 +156,7 @@ let
       '' + lib.optionalString (!stdenv.targetPlatform.isWasm) ''
         echo "--unwindlib=libunwind" >> $out/nix-support/cc-cflags
-      '' + lib.optionalString (!stdenv.targetPlatform.isWasm && stdenv.targetPlatform.useLLVM or false) ''
+      '' + lib.optionalString (!stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD && stdenv.targetPlatform.useLLVM or false) ''
```
@Mic92 I guess changes like this are adding the problem? But it also didn't work without. Hmm.

sternenseemann replied:
Yeah, this seems wrong as we are doing ` useLLVM`

Ericson2314 replied:
Well forcing `-lunwind` is a sledgehammer I am skeptical enough, and it was causing failures even if its absense was also causing failures. I think it would be good to figure out why this was needed in the first place.
